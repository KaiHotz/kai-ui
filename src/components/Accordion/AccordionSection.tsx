import React, { FC, ReactNode } from 'react';
import { useCollapse } from 'react-collapsed';
import cx from 'clsx';
import { FaChevronDown, FaChevronUp } from 'react-icons/fa6';

import { useAccordionSection } from './context';

export interface IAccordionSectionProps {
  children: ReactNode;
  title: string;
  className?: string;
  icon?: ReactNode;
}

export const AccordionSection: FC<IAccordionSectionProps> = ({ children, className, title, icon }) => {
  const { onClick, isExpanded, hasSimpleSeparator, testId } = useAccordionSection();
  const { getCollapseProps, getToggleProps } = useCollapse({ isExpanded, duration: 200, easing: 'linear' });
  const isExpandable = onClick !== undefined;
  const arrowIcon = isExpanded ? <FaChevronUp /> : <FaChevronDown />;

  return (
    <div
      className={cx(
        'ui-accordion__section',
        { 'ui-accordion__section--expanded': isExpanded, 'ui-accordion__section--expandable': isExpandable },
        className,
      )}
      data-testid={testId}
    >
      <div
        className={cx('ui-accordion__section-header', {
          'ui-accordion__section-header--simple-separator': hasSimpleSeparator,
          'ui-accordion__section-header--separator': !hasSimpleSeparator,
        })}
        {...getToggleProps({ onClick })}
        data-testid={`${testId}-header`}
      >
        <div className="ui-accordion__section-header-title" data-testid={`${testId}-title`}>
          {icon}
          <span className="ui-accordion__section-header-title-text">{title}</span>
        </div>
        {isExpandable && arrowIcon}
      </div>
      <section {...getCollapseProps()} className="ui-accordion__section-content">
        {children}
      </section>
    </div>
  );
};
