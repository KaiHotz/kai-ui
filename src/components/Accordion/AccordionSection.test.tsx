import React from 'react';
import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import { AccordionSection } from './AccordionSection';

const props = {
  title: 'accordionSection',
  className: 'Custom',
  children: 'dwadaw',
};

describe('<AccordionSection />', () => {
  it('should allow custom className', () => {
    render(<AccordionSection {...props} />);
    expect(screen.getByTestId('ui-accordion-section')).toHaveClass(props.className);
  });

  it('should show title', () => {
    render(<AccordionSection {...props} />);
    expect(screen.getByText(props.title)).toBeInTheDocument();
  });
  it('should show children', () => {
    render(<AccordionSection {...props} />);
    expect(screen.getByText(/dwadaw/i)).toBeInTheDocument();
  });
});
