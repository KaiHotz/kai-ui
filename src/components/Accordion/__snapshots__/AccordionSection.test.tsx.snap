// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<AccordionSection /> > should render component with an animation 1`] = `
<div
  class="ui-accordion__section ui-accordion__section--expandable Custom"
  data-testid="ui-accordion-section"
>
  <div
    aria-controls="react-collapsed-panel-:r0:"
    aria-expanded="false"
    class="ui-accordion__section-header ui-accordion__section-header--separator"
    data-testid="ui-accordion-section-header"
    id="react-collapsed-toggle-:r0:"
    role="button"
    tabindex="0"
  >
    <div
      class="ui-accordion__section-header-title"
      data-testid="ui-accordion-section-title"
    >
      <span
        class="ui-accordion__section-header-title-text"
      >
        accordionSection
      </span>
    </div>
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
      />
    </svg>
  </div>
  <section
    aria-hidden="true"
    aria-labelledby="react-collapsed-toggle-:r0:"
    class="ui-accordion__section-content"
    id="react-collapsed-panel-:r0:"
    role="region"
    style="box-sizing: border-box; display: none; height: 0px; overflow: hidden;"
  >
    dwadaw
  </section>
</div>
`;
